### PR TITLE
Add Query Result Reuse Support to Backend

### PR DESCRIPTION
**Pull request by kevinwcyu**
Adds backend support for query result reuse.

This depends on https://github.com/grafana/grafana-aws-sdk/pull/71

Fixes: https://github.com/grafana/oss-plugin-partnerships/issues/120

This uses the connection args to set the result reuse configuration. The connection args are per query. The connection args are passed into the `Connect` method for sync queries (this is done in sqlds https://github.com/grafana/sqlds/blob/f35c5a62f66c02aef63b260c1e0a25568ae698b6/datasource.go#L169), and are passed into the `GetAsyncDB` method for async queries (this will be done in grafana-aws-sdk in this PR https://github.com/grafana/grafana-aws-sdk/pull/71/files).

`Connect` and `GetAsyncDB` eventually call [`Settings.Apply`](https://github.com/grafana/grafana-aws-sdk/blob/fdb2a182c315dedb479698d4fdeb0f5af235e2b8/pkg/sql/datasource/datasource.go#L116) which is defined in `pkg/athena/models/settings.go` below. This applies the Result Reuse Configuration to the settings that is used in the `Execute` method in `pkg/athena/api/api.go`

The frontend will update the `AthenaQuery` type to include `resultReuseEnabled` and `resultReuseMaxAgeInMinutes` in `connectionArgs` here https://github.com/grafana/athena-datasource/blob/3b0518bf118910644c84082c03d222862f83133d/src/types.ts#L27-L31.

### CODE REVIEW

File: pkg/athena/datasource.go [hunk 1]
```diff
@@ -159,3 +160,20 @@ func (s *AthenaDatasource) Workgroups(ctx context.Context, options sqlds.Options
...
+	if queryArgs != nil {
+		err := json.Unmarshal(queryArgs, &args)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse query args: %w", err)
+		}
+	}
+
+	options := sqlds.Options{}
+	for k, v := range args {
+		options[k] = fmt.Sprintf("%v", v)
```
Comment (kevinwcyu) [inline pkg/athena/datasource.go:175]: Is it fine to convert `v` to a string in this way? I was previously doing this https://github.com/grafana/athena-datasource/pull/214/commits/7fdf12b664b06bd803fd8474c34fbfa52b953729.

**Review by idastambuk — APPROVED** [review]

File: pkg/athena/datasource.go (hunk 1 above)
Comment (fridgepoet) [reply pkg/athena/datasource.go:175]: In that commit you linked, was there a reason to get rid of the `toString` function?

So basically we're never going to know what we're unmarshalling? There is nothing we can describe about `args`'s types? We can only say it's interface{}?

My thought, but I ought to look at the rest of this PR and the part in grafana-aws-sdk to understand what will come out: %v is the most general string formatting verb, might not be in the format that the downstream consumers expect. https://gobyexample.com/string-formatting

Comment (iwysiu) [reply pkg/athena/datasource.go:175]: I don't think it's strictly wrong, but it's probably better practice to do the toString function with a switch statement for every type we expect and throw an error for unexpected types.

File: pkg/athena/api/api.go [hunk 2]
```diff
@@ -57,6 +57,12 @@ func (c *API) Execute(ctx context.Context, input *api.ExecuteQueryInput) (*api.E
...
 		WorkGroup: aws.String(c.settings.WorkGroup),
+		ResultReuseConfiguration: &athena.ResultReuseConfiguration{
+			ResultReuseByAgeConfiguration: &athena.ResultReuseByAgeConfiguration{
+				Enabled:         aws.Bool(c.settings.ResultReuseEnabled),
+				MaxAgeInMinutes: aws.Int64(c.settings.ResultReuseMaxAgeInMinutes),
+			},
+		},
 	}
 
 	if c.settings.OutputLocation != "" {
```
Comment (fridgepoet) [inline pkg/athena/api/api.go:68]: Just a note that I think that below here, after we receive the QueryExecutionId we'll be able to get additional information if result reuse was in place or not. See https://github.com/grafana/oss-plugin-partnerships/issues/127

`c.Client.GetQueryExecutionWithContext(ctx, &athena.GetQueryExecutionInput{QueryExecutionId: output.QueryExecutionId})`

**Review by fridgepoet — COMMENTED** [review]
Looks really cool!
I was able to test this out locally and use another AWS response API locally to see that result reuse can be enabled. (And it all worked out with go workspaces, no more go.mod modifications -- I also cloned and added the sqlds to my Athena workspace, so now it has athena-datasource, grafana-aws-sdk, and sqlds in go.work).

I did not yet get a chance to review the tests.

Would you mind checking out the two following points I'm wondering about?
* What happens when you try to set reuse result enabled = true when you use the athena engine v2?
* What happens if you set reuse result enabled = false and set a value for the maxAgeInMinutes?

File: pkg/athena/datasource.go [hunk 3]
```diff
@@ -159,3 +160,20 @@ func (s *AthenaDatasource) Workgroups(ctx context.Context, options sqlds.Options
...
+	}
+
+	return options, nil
+}
```
[the remark this answers is not in the document]
Comment (fridgepoet) [reply pkg/athena/datasource.go:179]: For the args like “database” etc in the map it might be nice to have this in constants which can also be used in the Apply function. It’ll be nice to know where these are used again. Let me know if this makes sense (writing on mobile sorry)

File: pkg/athena/api/api.go (hunk 2 above)
Comment (kevinwcyu) [reply pkg/athena/api/api.go:77]: i had to do the check in the `Status` method. when i did it immediately after getting the query execution id, the boolean that indicated whether it was reusing a query or not was always false for me. probably because it just started the query.

**Comment by fridgepoet** [comment]
Wondering if we should check if it is not equal to "engine version 2" instead of only allowing engine version 3. Would you check with Tolv about this? They commented in the design doc. 

> When reuse result enabled = false and set a value for the maxAgeInMinutes nothing happens.
Thanks for checking this. If we can capture this input in a unit test and show that we don't return an error for it, it explicitly documents we've thought about this case. What do you think? I think it is also fine if we don't include that.

Finally, we had an impromptu pairing session and just talked about tests -- couple of commits regarding that are captured here. https://github.com/grafana/athena-datasource/compare/kevinwcyu/120-test-fun

**Review by fridgepoet — APPROVED** [review]

**Comment by a-szegel** [comment]
Out of curiosity, what is the average time of a re-used query?

**Comment by a-szegel** [comment]
I don't know if this is possible (or already done), but from a user perspective, it would be nice to be able to make a dashboard variable to control whether QRR gets used or not (defaulting to QRR).

File: pkg/athena/datasource.go [hunk 4]
```diff
@@ -186,10 +185,20 @@ func parseArgs(queryArgs json.RawMessage) (sqlds.Options, error) {
 		}
 	}
 	options := sqlds.Options{}
-	options[models.Region] = args.Region
-	options[models.Catalog] = args.Catalog
-	options[models.Database] = args.Database
-	options[models.ResultReuseEnabled] = fmt.Sprintf("%t", args.ResultReuseEnabled)
-	options[models.ResultReuseMaxAgeInMinutes] = fmt.Sprintf("%d", args.ResultReuseMaxAgeInMinutes)
+	if args.Region != "" {
```
Comment (fridgepoet) [inline pkg/athena/datasource.go:188]: this kind of thing failed on an e2e test? we can check this in a unit test. (we can do this in a separate commit)

Comment (fridgepoet) [reply pkg/athena/datasource.go:188]: *separate PR
